File: Python/Problem_106.py
from math import comb

def catalan(k: int) -> int:
    return comb(2*k, k) // (k + 1)

def needed_tests(n: int) -> int:
    total = 0
    for k in range(2, n//2 + 1):
        ways_pick = comb(n, 2*k)
        splits = comb(2*k, k)
        noncross = catalan(k)
        need = (splits - 2*noncross) // 2
        total += ways_pick * need
    return total

if __name__ == "__main__":
    print(needed_tests(12))  # 21384

File: Java/Problem_106.java
public class Main {
    static long comb(int n, int r) {
        if (r < 0 || r > n) return 0;
        r = Math.min(r, n - r);
        long res = 1;
        for (int i = 1; i <= r; i++) {
            res = res * (n - r + i) / i;
        }
        return res;
    }

    static long catalan(int k) {
        return comb(2 * k, k) / (k + 1);
    }

    static long neededTests(int n) {
        long total = 0;
        for (int k = 2; k <= n / 2; k++) {
            long waysPick = comb(n, 2 * k);
            long splits = comb(2 * k, k);
            long noncross = catalan(k);
            long need = (splits - 2 * noncross) / 2;
            total += waysPick * need;
        }
        return total;
    }

    public static void main(String[] args) {
        System.out.println(neededTests(12)); // 21384
    }
}
